### DISCUSSION TITLE
how to get headers/body/query request parameter  at the same time with multiple extractors

### DISCUSSION THREAD

**Discussion by ggslayer** [tokio-rs/axum#320] [Q&A] [answered]
how to get headers/body/query request parameter  at the same time with multiple extractors?

example request:
http://xx.xx.xx.xx/api/set_book_content?sec_id=xx&sec_key=xx
with header: header_name_xx=header_val_xx
with body: this is book's content text ... long text
with cookie: cookie_name_xx=cookie_value_xx

in my handler, i wan to write like this:
async fn get_test(body_text:String, headers:Headers<HashMap<String,String>>,cookies:Cookies<HashMap<String,String>>,query_params:Form<my_query>) -> Json<my_response>{}

my propose is：i want get the four data in the simplist way.  please tell me the simplist way, thank you!!

**Comment by davidpdrsn** [comment, marked as answer]
You can apply multiple extractors to a function so it’s just a matter of using the right types:

- Body can be extracted with `String`, `Bytes`, or several others in the extract module.
- You can get cookies with `TypedHeader<headers::Cookie>`.
- `HeaderMap` gives you the headers. Apply this after cookies since it will remove the headers from the request.
- `extract::Query` or `RawQuery` gives you the query.

**Reply by ggslayer** [reply]
thank you very much

**Reply by ggslayer** [reply]
hi, when i do that follow you , i got error: "Header of type `cookie` was missing".

this is my func define:
async fn get_test(body:Form<TestReqData>, cookies:TypedHeader<headers::Cookie>, headers:HeaderMap, query_param:Query<TestQueryData> ) -> Json<TestResponseData>
{
  ...
}

**Reply by davidpdrsn** [reply]
Does the request actually contain a cookie header? If it’s optional you must use `Option<TypedHeader<Cookie>>`.
